packages/shell-api/src/stream-processor.ts

```diff
@@ -0,0 +1,121 @@
+import { promisify } from 'util';
+
+import type { Document } from '@mongosh/service-provider-core';
+
+import type Mongo from './mongo';
+import { asPrintable } from './enums';
+import {
+  ShellApiWithMongoClass,
+  returnsPromise,
+  shellApiClassDefault,
+} from './decorators';
+
+import type { Streams } from './streams';
+
+const sleep = promisify(setTimeout);
+
+@shellApiClassDefault
+export default class StreamProcessor extends ShellApiWithMongoClass {
+  constructor(public _streams: Streams, public name: string) {
+    super();
+  }
+
+  get _mongo(): Mongo {
+    return this._streams._mongo;
+  }
+
+  [asPrintable]() {
+    return `Atlas Stream Processor: ${this.name}`;
+  }
+
+  @returnsPromise
+  async start() {
+    const result = await this._streams._runCommand({
+      startStreamProcessor: 1,
+      name: this.name,
+    });
+    return result;
+  }
+
+  @returnsPromise
+  async stop() {
+    const result = await this._streams._runCommand({
+      stopStreamProcessor: 1,
+      name: this.name,
+    });
+    return result;
+  }
+
+  @returnsPromise
+  async drop() {
+    const result = await this._streams._runCommand({
+      dropStreamProcessor: 1,
+      name: this.name,
+    });
+    return result;
+  }
+
+  @returnsPromise
+  async stats(options: Document = {}) {
+    return this._streams._runCommand({
+      getStreamProcessorStats: 1,
+      name: this.name,
+      ...options,
+    });
+  }
+
+  @returnsPromise
+  async sample(options: Document = {}) {
+    const r = await this._streams._runCommand({
+      startSampleStreamProcessor: 1,
+      name: this.name,
+      ...options,
+    });
+
+    if (r.ok !== 1) {
+      return r;
+    }
+
+    return this._sampleFrom(r.cursorId as number);
+  }
+
+  @returnsPromise
```

Comment:
I may be misremembering, but if this is a private method that is not meant to be used by users directly you shouldn't need to annotate it.

`@returnsPromise` marks the method such that when a user uses it in the REPL it gets rewritten to be "auto-awaited" and it appears as if it would be synchronous, for example, by marking`findOne` with `@returnsPromise` users are able to do this:

```
db.coll.findOne().x
```

Reply:
Yes, I added this because it's available and can be called with `sp.someStream._sampleFrom()` even though its name indicates this is private and should not be used directly. I don't mind removing it.

Reply:
Ah ok, there should be no harm into keeping it either 🤷‍♂️